src/tree/AATree.scala

```diff
@@ -0,0 +1,220 @@
+ * red-black tree. While an AA tree makes more rotations than a red-black tree,
+ * the simpler algorithms tend to be faster, and all of this balances out to
+ * result in similar performance. A red-black tree is more consistent in its
+ * performance than an AA tree, but an AA tree tends to be flatter, which
+ * results in slightly faster search times
+ *
+ * Insert - O(log n)
+ * Lookup - O(log n)  
+ * Remove - O(log n)
+ */
+
+object AATree {
+
+  sealed abstract class AATree[+T: Ordering] {
+    val value: T
+    val level: Int
+    val left: AATree[T]
+    val right: AATree[T]
+
+    def member[B >: T: Ordering](a: B): Boolean = this match {
+      case Empty                                                => false
+      case Fork(v, _, l, r) if a == v                           => true
+      case Fork(v, _, l, r) if implicitly[Ordering[B]].lt(a, v) => l.member(v)
+      case Fork(v, _, l, r)                                     => r.member(v)
+    }
+
+    def +[B >: T: Ordering](a: B): AATree[B] = this match {
+      case Empty                                                 =>
+        Fork(a, 1, Empty, Empty)
+      case Fork(v, lv, l, r) if implicitly[Ordering[B]].lt(a, v) => 
+        Fork(v, lv, l.insert(a), r).rebalanceTree
+      case Fork(v, lv, l, r) if implicitly[Ordering[B]].gt(a, v) => 
+        Fork(v, lv, l, r.insert(a)).rebalanceTree
+      case _                                                     =>
+        this
+    }
+
+    def insert[B >: T: Ordering](a: B): AATree[B] = insert(a)
+    def add[B >: T: Ordering](a: B): AATree[B] = insert(a)
+
+    def insertAll[B >: T: Ordering](as: List[B]): AATree[B] = {
+      as.foldLeft(this: AATree[B])((z: AATree[B], y: B) => z + y)
+    }
+
+    def merge[B >: T: Ordering](that: AATree[B]): AATree[B] = 
+      if(this == Empty) that
+      else if(that == Empty) this
+      else {
+        insertAll(that.toList)
+      }
+
+    def ++[B >: T: Ordering](tree: AATree[B]): AATree[B] = merge(tree)
+
+    // Let's make our tree an instance of functor, and monad
+    def map[B: Ordering](f: T => B): AATree[B] = this match {
+      case Empty            => Empty
+      case Fork(v, _, l, r) =>
+        val mappedL = l.map(f)
+        val mappedR = r.map(f)
+        mappedL.merge(mappedR) + f(v)
+    }
+
+    def flatMap[B: Ordering](f: T => AATree[B]): AATree[B] = this match {
+      case Empty            => Empty
+      case Fork(v, _, l, r) =>
+        val mappedL = l.flatMap(f)
+        val mappedR = r.flatMap(f)
+        mappedL.merge(mappedR).merge(f(v))
+    }
+
+
+
+    def foreach(f: T => Unit): Unit = toList.foreach(f)
+
+    def filter(p: T => Boolean): AATree[T] = Empty.insertAll(toList.filter(p))
+     
+
+    def fold[B: Ordering](f: (T, B, B) => B, z: B): B = this match {
+      case Empty            => z
+      case Fork(v, _, l, r) =>
+        f(v, l.fold(f, z), r.fold(f, z))
+    }
+
+    def delete[B >: T: Ordering](a: B): AATree[B] = this match {
```

Comment:
It should probably be named `remove` instead. `insert`/`remove` but `create`/`delete`.